api/src/api/diag.ts

```diff
@@ -101,7 +101,7 @@ export class DiagAPI implements DiagLogger, DiagLoggerApi {
         );
       }
 
-      return registerGlobal('diag', newLogger, self, true);
+      return registerGlobal('diag', newLogger, newLogger, true);
```

Comment:
Actually, the original implementation (passing `self`) is technically the correct way -- but this is a chicken / egg situation.

So if the existing logger is effectively a Noop then it's silent, but with the new code it the "new" logger is a Noop then you also get the same situation. 

My preference is to keep the "current" (self) implementation so that the application that registers the logger is the logger that receives the failure and not some logger that an instrumentation may add after the fact.

Reply:
I might be misunderstanding and/or have my patch wrong, but I'm confused by this part of your feedback:

> My preference is to keep the "current" (self) implementation so that the application that registers the logger is the logger that receives the failure and not some logger that an instrumentation may add after the fact.

The scenario I was trying to cover:
1. First, `diag.setLogger(...)` is set using an older API version (e.g. 1.5.0), perhaps by some dependency of my app that I'm not aware is using OTel.
2. Then, my application is using a newer API version (e.g. 1.7.0) and attempts to set a diag logger and use it.

In this scenario, my usage of diag logging (e.g. `api.diag.info(...)`) will silently not log **and there will be no error message to either the old or the new diag logger about the `setLogger` registration having failed.**
In this scenario, "the application that registers the logger" does not receive the failure.

(It has been a while, so I don't recall exactly how I was hitting this. I think I was probably working on some issue in the contrib repo and accidentally got a repo state where there were two versions of `@opentelemetry/api` active. I hit this case and it look a while to figure out why my registered diag logger was completely silent.)

Reply:
As an alternative, we can similarly early check the registered API version, like when overriding a diag logger https://github.com/open-telemetry/opentelemetry-js/blob/main/api/src/api/diag.ts#L90-L102.

```js
      const oldVersion = getGlobalVersion();
      if (oldVersion && oldVersion !== VERSION) {
        const stack = new Error().stack ?? '<failed to generate stacktrace>';
        const message = `Current logger (${oldVersion}) can not be overridden by version ${VERSION} ${stack}`;
        oldLogger.warn(message);
        newLogger.warn(message);
      }
```